Support saving full model when performing LoRA training

hendrydong opened the pull request:
Run

run_finetune_with_lora_save_aggregated_weights.sh and train_merged_lora.sh

to perform two stage LoRA training

research4pan reviewed (changes requested):
### `requirements.txt`

- line 3: commit id may need to be provided, since the main branch of peft keeps changing, which may affect the functionality of our package.

### `scripts/run_finetune_with_lora.sh`

- [Question] line 32: I am wondering if zero3 optimization can be supported.

### `scripts/run_finetune_with_lora_save_aggregated_wieghts.sh`

- [Style] line 3-4: better removed

### `scripts/export_llama_state_dict_checkpoint.py`

- [Question] Does our users need this script? If so, how to use it?

### `scripts/train_merged_lora.sh`

- [Style] line 3-4: better removed
- [Question] Can we make lora's default behavior as `--save_aggregated_lora 1`, so that it can have the same interface as `finetune.sh` and don't have to create a new script (which can be a burden for our users).
- [Question] What's the difference between `scripts/run_finetune_with_lora_save_aggregated_weights.sh` and `scripts/train_merged_lora.sh`?

### `scripts/lmflow/models/hf_decoder_model.py`

- [Question] line 271-272: I am wondering if the removal of those two lines is safe under all scenarios, like when we use lora but don't turn on `--save_aggregated_lora`.
- [Question] line 202: I am wondering if this line is necessary.
- [Style] line 241: better removed
- [Architecture] line 237-244: better encapsulate this in a new model method `hf_decoder_model.save()`. We don't want the pipeline to know and handle the details of model management, since it other contributors may not be familiar with that part.

hendrydong commented:
> @hendrydong Thanks! This feature is quite important and requires a lot of efforts to implement. Thanks for this meaningful contribution! Before merging into the main branch, we may need to fix some minor problems first, so that the main branch is still stable and friendly to other users.
> 
> ### `requirements.txt`
> * line 3: commit id may need to be provided, since the main branch of peft keeps changing, which may affect the functionality of our package.
> 

I have edit the requirements.txt. BTW, the previous branch specification is incorrect and I have correct them.

> ### `scripts/run_finetune_with_lora.sh`
> * [Question] line 32: I am wondering if zero3 optimization can be supported.
> 

Yes. I have tested.

> ### `scripts/run_finetune_with_lora_save_aggregated_wieghts.sh`
> * [Style] line 3-4: better removed
> 

Done.

> ### `scripts/export_llama_state_dict_checkpoint.py`
> * [Question] I am wondering if our users need this script. If so, I am wondering how it is used. We may need a README section in that case.
> 

Added.


> ### `scripts/train_merged_lora.sh`
> * [Style] line 3-4: better removed
> * [Question] Can we make lora's default behavior as `--save_aggregated_lora 1`, so that it can have the same interface as `finetune.sh` and don't have to create this new script (which can be a burden for our users). Better if it can support seamingless multi-stage training with LoRA and non-LoRA models.

In fact, the key issue is that we do not need the aggregated one by default. The aggregated version is reserved for those who need the full model.

> * [Question] I am wondering the if there is any difference between `scripts/run_finetune_with_lora_save_aggregated_weights.sh` and `scripts/train_merged_lora.sh`.
>

I have removed train_merged_lora.
 
> ### `scripts/lmflow/models/hf_decoder_model.py`
> * [Question] line 271-272: I am wondering if the removal of those two lines is safe under all scenarios, like when we use lora but don't turn on `--save_aggregated_lora`.

It is safe. I know why I wrote this two line but we remove the usage now.

> * [Question] line 202: I am wondering if this line is necessary.

Removed (I think this is the finetuner.py).
> * [Style] line 241: better removed

Done.

> * [Architecture] line 237-244: better encapsulate this in a new model method `hf_decoder_model.save()`. We don't want the pipeline to know and handle the details of model management, since other contributors may not be familiar with that part.

Done.


Hi Rui,

I have upload another version. Can you check it again? Thank you very much.

research4pan reviewed (approved):
Thanks! The modifications look good to me. Some minor problems that may need to be modified later, but no need to rush in this pull request. Thanks for @hendrydong for this meaningful contribution!

### `scripts/run_finetune_with_lora.sh`

- [Style] line 30: can we reserve the zero3 option, so that users who can run their scripts before doesn't have to experience differences?

### `scripts/run_finetune_with_lora_save_aggregated_weights.sh`

- [Style] Maybe we can write the whole 2-stage training process here so that users don't have to run this script multiple times.

### `src/lmflow/models/hf_decoder_model.py`

- [Style] line 161, 224, 271: extra empty line
- [Architecture] line 406-410: it is better to move `merge_lora_weights` into `save` option, or make it a private method. Since it is normally not used under other scenarios.

### `src/lmflow/pipeline/finetuner.py`

- [Style] line 202: better remove leading spaces.
- [Architecture] line 237-239: better invoke `model.merge_lora_weight` in `model.save` and hide this method, since in the future, other models may not support LoRA, such as customized models, but we still want to reuse this `finetuner` to tune those models.
